### PR TITLE
Increase test coverage, coerce inputs to string and fix encoding inconsistencies

### PR DESCRIPTION
**Pull request by mudge**
Where possible, coerce inputs to strings with StringValue rather than raising a TypeError. This particularly impacts the relatively recent RE2::Set API which was excessively strict about its arguments.

Add test coverage to all parts of the API, better covering edge cases including how encoding is handled based on the encoding of the RE2 pattern.

### CODE REVIEW

File: spec/re2_spec.rb [hunk 1]
```diff
@@ -15,29 +15,68 @@
...
+    end
+
+    it "returns ISO-8859-1 strings if the pattern is not UTF-8" do
+      original = "Foo".encode("Shift_JIS")
+      replacement = RE2.Replace(original, RE2("oo", :utf8 => false), "ah")
+
+      expect(replacement.encoding).to eq(Encoding::ISO_8859_1)
+    end
+
+    it "returns strings in the encoding of the given String pattern" do
```
Comment (mudge) [inline spec/re2_spec.rb:60]: It’s worth noting this probably isn’t the right behaviour: the pattern will be compiled into an `RE2` but the input will be interpreted as UTF-8 by default so any result will also be UTF-8. (Not to mention this will probably result in undefined behaviour if passing non-UTF-8 data.)

File: spec/re2_spec.rb [hunk 2]
```diff
@@ -53,23 +92,78 @@
...
+    end
+
+    it "returns ISO-8859-1 strings if the pattern is not UTF-8" do
+      original = "Foo".encode("Shift_JIS")
+      replacement = RE2.GlobalReplace(original, RE2("oo", :utf8 => false), "ah")
+
+      expect(replacement.encoding).to eq(Encoding::ISO_8859_1)
+    end
+
+    it "returns strings in the encoding of the given String pattern" do
```
Comment (mudge) [inline spec/re2_spec.rb:137]: As above, this should probably be UTF-8 as that is what RE2 will use.

**Review by mudge — COMMENTED** [review]
This will close https://github.com/mudge/re2/issues/74 and https://github.com/mudge/re2/issues/73.

File: spec/re2/scanner_spec.rb [hunk 3]
```diff
@@ -21,9 +21,10 @@
   end
 
   describe "#scan" do
-    it "returns the next array of matches" do
+    it "returns the next array of matches", :aggregate_failures do
       r = RE2::Regexp.new('(\w+)')
       scanner = r.scan("It is a truth universally acknowledged")
+
       expect(scanner.scan).to eq(["It"])
```
Comment (mudge) [inline spec/re2/scanner_spec.rb:28]: We should also assert the encoding behaviour here.

File: ext/re2/re2.cc [hunk 4]
```diff
@@ -1418,8 +1410,8 @@ static VALUE re2_Replace(VALUE self, VALUE str, VALUE pattern,
     RE2::Replace(&str_as_string, StringValuePtr(pattern),
         StringValuePtr(rewrite));
 
-    return ENCODED_STR_NEW2(str_as_string.data(), str_as_string.size(),
-        pattern);
+    return ENCODED_STR_NEW(str_as_string.data(), str_as_string.size(),
```
Comment (mudge) [inline ext/re2/re2.cc:1413]: Given this change, we should document how RE2 only supports UTF-8 and ISO-8859-1 and anything else is undefined.

File: ext/re2/re2.cc [hunk 5]
```diff
@@ -1418,15 +1461,19 @@ static VALUE re2_Replace(VALUE self, VALUE str, VALUE pattern,
...
+        "UTF-8");
```
Comment (mudge) [inline ext/re2/re2.cc:1465]: @stanhu will this (and the change to `GlobalReplace` below) affect GitLab or are you only using UTF-8 strings with re2?

Comment (stanhu) [reply ext/re2/re2.cc:1465]: As far as I can tell, GitLab is only using UTF-8 strings with re2.
